File: lang/c/src/vpkppc/format/GCF.cpp
#include <vpkppc/format/GCF.h>

#include <sourceppc/Helpers.h>

using namespace sourceppc;
using namespace vpkpp;

const char* VPKPP_GCF_EXTENSION = GCF_EXTENSION.data();

SOURCEPP_API vpkpp_pack_file_handle_t vpkpp_gcf_open(const char* path, vpkpp_entry_callback_t callback, vpkpp_pack_file_open_property_request_t requestProperty) {
	SOURCEPP_EARLY_RETURN_VAL(path, nullptr);

	auto packFile = GCF::open(path, callback ? [callback](const std::string& entryPath, const Entry& entry) {
		callback(entryPath.c_str(), const_cast<Entry*>(&entry));
	} : static_cast<PackFile::EntryCallback>(nullptr), requestProperty ? [requestProperty](PackFile* packFile_, PackFile::OpenProperty property) {
		return convert::fromBuffer<std::byte>(requestProperty(packFile_, static_cast<vpkpp_pack_file_open_property_e>(property)));
	} : static_cast<PackFile::OpenPropertyRequest>(nullptr));
	if (!packFile) {
		return nullptr;
	}
	return packFile.release();
}

SOURCEPP_API uint32_t vpkpp_gcf_get_version(vpkpp_pack_file_handle_t handle) {
	SOURCEPP_EARLY_RETURN_VAL(handle, 0);

	const auto* gcf = dynamic_cast<GCF*>(convert::handle<PackFile>(handle));
	SOURCEPP_EARLY_RETURN_VAL(gcf, 0);

	return gcf->getVersion();
}

SOURCEPP_API uint32_t vpkpp_gcf_get_appid(vpkpp_pack_file_handle_t handle) {
	SOURCEPP_EARLY_RETURN_VAL(handle, 0);

	const auto* gcf = dynamic_cast<GCF*>(convert::handle<PackFile>(handle));
	SOURCEPP_EARLY_RETURN_VAL(gcf, 0);

	return gcf->getAppID();
}

SOURCEPP_API uint32_t vpkpp_gcf_get_app_version(vpkpp_pack_file_handle_t handle) {
	SOURCEPP_EARLY_RETURN_VAL(handle, 0);

	const auto* gcf = dynamic_cast<GCF*>(convert::handle<PackFile>(handle));
	SOURCEPP_EARLY_RETURN_VAL(gcf, 0);

	return gcf->getAppVersion();
}

File: lang/csharp/src/vpkpp/Format/GCF.cs
using System;

namespace sourcepp.vpkpp.Format;

using EntryCallback = Action<string, Entry>;
using OpenPropertyRequest = Func<PackFile, OpenProperty, byte[]>;

public class GCF : PackFile
{
	protected GCF(nint handle, bool managed = true) : base(handle, managed)
	{
	}

	public new static GCF? Open(string path, EntryCallback? callback = null, OpenPropertyRequest? requestProperty = null)
	{
		var handle = DLL.vpkpp_gcf_open(path, callback is not null ? (entryPath, entry) =>
		{
			callback(entryPath, new Entry(entry, false));
		} : null, requestProperty is not null ? (handle, property) =>
		{
			var data = requestProperty(new PackFile(handle, false), property);
			return sourcepp.DLL.sourcepp_buffer_new((ulong) data.Length);
		} : null);
		return handle == nint.Zero ? null : new GCF(handle);
	}

	public uint Version
	{
		get
		{
			ThrowIfDisposed();
			return DLL.vpkpp_gcf_get_version(Handle);
		}
	}

	public uint AppId
	{
		get
		{
			ThrowIfDisposed();
			return DLL.vpkpp_gcf_get_appid(Handle);
		}
	}

	public uint AppVersion
	{
		get
		{
			ThrowIfDisposed();
			return DLL.vpkpp_gcf_get_app_version(Handle);
		}
	}
}
